src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,926 @@
+        size_t next_child = 0;
+    };
+
+    std::stack<Frame> stack;
+    std::unordered_set<const ActionsDAG::Node *> visited;
+    for (const auto & node : dag.getNodes())
+    {
+        if (visited.contains(&node))
+            continue;
+
+        stack.push({&node});
+        visited.insert(&node);
+        while (!stack.empty())
+        {
+            auto & frame = stack.top();
+            for (; frame.next_child < frame.node->children.size(); ++frame.next_child)
+                if (!visited.contains(frame.node->children[frame.next_child]))
+                    break;
+
+            if (frame.next_child < frame.node->children.size())
+            {
+                const auto * child = frame.node->children[frame.next_child];
+                visited.insert(child);
+                stack.push({child});
+                ++frame.next_child;
+            }
+            else
+            {
+                /// Ignore constants here, will check them separately
+                if (!frame.node->column)
+                {
+                    if (frame.node->type == ActionsDAG::ActionType::ALIAS)
+                    {
+                        if (fixed_columns.contains(frame.node->children.at(0)))
+                            fixed_columns.insert(frame.node);
+                    }
+                    else if (frame.node->type == ActionsDAG::ActionType::FUNCTION)
+                    {
+                        if (frame.node->function_base->isDeterministicInScopeOfQuery())
+                        {
+                            //std::cerr << "*** enreachFixedColumns check " << frame.node->result_name << std::endl;
+                            bool all_args_fixed_or_const = true;
+                            for (const auto * child : frame.node->children)
+                            {
+                                if (!child->column && !fixed_columns.contains(child))
+                                {
+                                    //std::cerr << "*** enreachFixedColumns fail " << child->result_name <<  ' ' << static_cast<const void *>(child) << std::endl;
+                                    all_args_fixed_or_const = false;
+                                }
+                            }
+
+                            if (all_args_fixed_or_const)
+                            {
+                                //std::cerr << "*** enreachFixedColumns add " << frame.node->result_name << ' ' << static_cast<const void *>(frame.node) << std::endl;
+                                fixed_columns.insert(frame.node);
+                            }
+                        }
+                    }
+                }
+
+                stack.pop();
+            }
+        }
+    }
+}
+
+/// This structure stores a node mapping from one DAG to another.
+/// The rule is following:
+/// * Input nodes are mapped by name.
+/// * Function is mapped to function if all children are mapped and function names are same.
```

Comment:
Does it mean that function is mapped to itself?

Reply:
Yes, but from different DAGs.
I mean, this is the idea - to find equal calculations if DAGs